### DISCUSSION TITLE
Where to access stdin to provide input?

### DISCUSSION THREAD

**Discussion by alycklama** [scalameta/metals#3486] [Q&A]
I'm testing nvim-metals with a simple hello world console app. So far I have not figured out how to access stdin to provide the input.

Looking at DAP, I saw that each debugging adapter can potentially launch the debugee in an integrated terminal or
an external terminal (see section **dap-terminal**), but I have not seen this in the nvim-metals codebase:
https://github.com/mfussenegger/nvim-dap/blob/master/doc/dap.txt

Any idea how I can access stdin to provide input?

**Comment by ckipp01** [comment]
Hey @alycklama, I transfered this over to the main Metals repo since this isn't actually unique to `nvim-metals` but something that has also been asked before about the other editors. Currently there is no way to do this with Metals.

**Reply by Dragon1573** [reply]
@alycklama, I met the same problem as you. 

**Metals** is currently running/debugging projects via built-in _Debugging Console_. This console is READ-ONLY and does NOT support user input.

I'm writing Java projects in VSCode too. The **Java Extension Pack for VSCode** will compile the project in background and launch the application in integrated terminal, both in "Run" and "Debug" mode. In that situation, we can easily interact with our application through Standard I/O via integrated terminal.

If **Metals** extension provides an option for us to switch running/debugging environment on our own, it would be great!

## My Environment

- Windows 10 Home China
- Microsoft Build OpenJDK v17.0.1 with Hotspot (via `winget`)
- Scala SDK v2.13.8 (via `winget`)
- Simple Build Tool v1.6.1 (via `winget`)
- Microsoft Visual Studio Code v1.63.2 (system install) (via `winget`)
- Metals Extension for VSCode v1.11.0

## My walkaround

If the project HAVE TO interact with users through `scala.io.StdIn`, you can launch the integrated terminal on your own and invoke `sbt run` to compile/launch the application.

**Reply by tgodzik** [reply]
An alternative would be to run in a separate process with the agent settings and then to attach to that process manually, though that is ont perfect. We could do that automatically for you and I suspect this is how java language server does it.
